`$` now works for  unsigned intergers with `nim js`

timotheecour opened the pull request.

timotheecour committed:
$(unsigned) now works for js

metagn commented on lib/system/dollars.nim:
```diff
@@ -1,8 +1,25 @@
   ## spelling `toString`:idx:.
 
+when defined(js):
+  when (NimMajor, NimMinor, NimPatch) >= (1, 3, 3):
```
Shouldn't this be `>= (1, 3, 2)`? If the version is updated from 1.3.1 to 1.3.3 and `>= (1, 3)` was the norm before then `since`s should also be even number patches.

timotheecour replied:
i thought about this but:
* see https://github.com/nim-lang/Nim/blob/af9c85270198455c7ba218fae919e79060a8960a/compiler/docgen.nim#L736 which uses NimPatch as indicator
* it reserves a possibility to have "unstable releases" on `1.(2x+1).(2y)` branches which could be a neat idea

narimiran replied:
What exactly is Nim v1.3.3, @timotheecour?

(Spoiler: it doesn't exist nor will it exist)

timotheecour replied:
that was already reverted from this PR and is discussed elsewhere

Araq commented:
This patch would be better without introducing your own ideas about how semver should work.

timotheecour committed:
fixup

timotheecour committed:
undo NimPatch bump for now

timotheecour commented:
PTAL

> This patch would be better without introducing your own ideas about how semver should work.

fine, I've reverted NimPatch bump from this PR, that aspect can be discussed separately in https://github.com/nim-lang/Nim/pull/14124

narimiran commented on lib/system.nim:
```diff
@@ -54,6 +54,16 @@ type
 
 include "system/basic_types"
 
+const
```
Why this moving around?

Clyybber replied:
Because otherwise thise symbols are not available in dollars.nim afaict.
IMO NimVersion should be moved too then.

narimiran replied:
You mean the stuff that was *literally* two lines above, so now this version stuff is *literally* 2000 lines apart? :facepalm:

timotheecour replied:
> IMO NimVersion should be moved too then.

it can't; `NimVersion` relies on what's defined in dollars, so these need to be split apart anyways.
Putting it to the top right after `int` is defined enables using branching on Nim version below that. The fact that it's 2000 lines apart makes only a cosmetic difference.

narimiran commented on lib/system.nim:
```diff
@@ -54,6 +54,16 @@ type
+
+  NimMinor* {.intdefine.}: int = 3
+    ## is the minor number of Nim's version.
+    # Odd for devel, even for releases.
```
Who is this comment for?

timotheecour replied:
For anyone looking at the code (just like any regular non doc comment).

timotheecour replied:
turned it into a regular doc comment

narimiran commented:
All changes in `system.nim` could be (and should be, IMO) reverted. They have nothing to do with the stuff this PR tries to fix.

EDIT: Ok, `dollars` need version info, but no need to move it 2000 lines above.

Clyybber commented on lib/system/dollars.nim:
```diff
@@ -1,8 +1,25 @@
   ## spelling `toString`:idx:.
 
+when defined(js):
+  when (NimMajor, NimMinor) >= (1, 3):
```
Use since please

timotheecour replied:
done

timotheecour committed:
use since; turn `Odd for devel, even for releases.` into a doc comment

timotheecour committed:
move NimMajor+friends closer to NimVersion according as per reviewer feedback

timotheecour commented:
> All changes in system.nim could be (and should be, IMO) reverted. They have nothing to do with the stuff this PR tries to fix.


"all changes" is litterally only moving the nim version tuple, and is a required change; the reason I had moved it to the top is that it's a commonly used symbol for versioning, and it makes sense to define it as early as possible, instead of having to move it each time. It already has to be separated from NimVersion anyways, so 2000 lines vs 4 lines makes only a cosmetic difference. 

but I've now moved it right above system/dollars not far from Nimversion for the sake of moving on.

timotheecour committed:
fixup
